PR 2 of n: Add first ML notebook on clustering and heatmaps (#259)
AlexsLemonade/training-modules: machine-learning/01-openpbta_heatmap.Rmd

---
title: "OpenPBTA: Create a heatmap"
output:   
  html_notebook: 
    toc: true
    toc_float: true
author: CCDL for ALSF
date: 2020
---

In this notebook, we cluster RNA-seq data from the Open Pediatric Brain Tumor Atlas (OpenPBTA) project and create a heatmap.
OpenPBTA is a collaborative project organized by the CCDL and the Center for Data-Driven Discovery in Biomedicine (D3b) at the Children's Hospital of Philadelphia conducted openly on GitHub.

You can read more about the project [here](https://github.com/alexslemonade/openpbta-analysis/#openpbta-analysis).

We've downloaded the publicly available data from the project and performed a variance stabilizing transformation on the stranded RNA-seq dataset.
(That code is available in `setup`!)

We'll use a package called [`ComplexHeatmap`](https://www.bioconductor.org/packages/release/bioc/html/ComplexHeatmap.html) to make our heatmap.
This package also will perform clustering as part of the heatmap.

> It is highly flexible; see the [`ComplexHeatmap` Complete Reference](https://jokergoo.github.io/ComplexHeatmap-reference/book/) for more information.

Reviewer: Warn also that it is opinionated? You have to get the data structures *just right*

## Set up

### Libraries

```{r load_library}
# We'll be doing a bit of data wrangling with the Tidyverse at the outset
library(tidyverse)
# ComplexHeatmap is the package we'll use for making a heatmap
# It will do the hierarchical clustering for us as well
library(ComplexHeatmap)
```

### Directories and files

We store the data we'll use in this notebook in `data/open-pbta/processed`. 

```{r directories, live = TRUE}
data_dir <- file.path("data", "open-pbta", "processed")

# We'll store the heatmap in plots - create directory if it doesn't exist yet
plots_dir <- "plots"
if (!dir.exists(plots_dir)) {
  dir.create(plots_dir, recursive = TRUE)
}
```

#### Input files

```{r input_files}
histologies_file <- file.path(data_dir, "pbta-histologies-stranded-rnaseq.tsv")
rnaseq_file <- file.path(data_dir, "pbta-vst-stranded.tsv") 
```

#### Output files

```{r png_output, live = TRUE}
heatmap_file <- file.path(plots_dir, 

>                           "common_histologies_high_variance_heatmap.png")

Author: Only one way to find out 😈

### Annotation

First, we'll set up the metadata that we want to use to label samples in the heatmap.
In `ComplexHeatmap` terminology, this is called annotation, or `HeatmapAnnotation`, specifically.

```{r sample_annotation_df}
sample_annotation_df <- filtered_histologies_df %>%
  # Select only the columns that we'll use
  select(Kids_First_Biospecimen_ID,
         short_histology,
         composition,
         tumor_descriptor)

# Let's examine these columns
sample_annotation_df
```

Right now, `sample_annotation_df` is a tibble because it's a filtered version of something we read in with `readr::read_tsv()`.
Usually this is pretty much equivalent to a data frame.

```{r sample_annotation_class}
class(sample_annotation_df)
```

BUT `ComplexHeatmap` expects a data frame, so sometimes it will behave in unexpected ways when you give it a tibble! 
Let's make a data frame and make the sample identifiers the rownames.

```{r annotation_as_df, live = TRUE}
sample_annotation_df <- sample_annotation_df %>%
  as.data.frame() %>%
  tibble::column_to_rownames("Kids_First_Biospecimen_ID")
```

To specify the colors in our annotation bar, we need to create a list of named vectors.

> The names of the list need to exactly match the column names in `sample_annotation_df` and the names in each vector need to exact match the values in those columns.

Reviewer: ```suggestion
The names of the list need to *exactly* match the column names in `sample_annotation_df` and the names in each vector need to *exactly* match the values in those columns.
```

```{r color_palettes}
# Palette recommended for those with color vision deficencies - we'll use that 
# for histologies
histology_colors <- colorblindr::palette_OkabeIto[2:5]
sample_annotation_colors <- list(
  short_histology = c(
    "LGAT" = histology_colors[1],
    "Ependymoma" = histology_colors[2],
    "HGAT" = histology_colors[3],
    "Medulloblastoma" = histology_colors[4]
  ),
  composition = c(
    "Solid Tissue" = "#A0A0A0",  # light grey
    "Derived Cell Line" = "#000000"  # black
  ),
  tumor_descriptor = c(
    "Initial CNS Tumor" = "#3333FF",
    "Progressive" = "#FFFF99",
    "Recurrence" = "#CCCCFF",
    "Second Malignancy" = "#000033",
    "Unavailable" = "#FFFFFF"  # white for missing data
  )
)
```

> We need to create a special type of object with `HeatmapAnnotation`.

Reviewer: ```suggestion
We need to create a special type of object with `HeatmapAnnotation` using the annotation data frame and the list of color vectors.
```

```{r create_annotation, live = TRUE}
column_annotation <- HeatmapAnnotation(
  df = sample_annotation_df,
  col = sample_annotation_colors
)
```

### Values for display

We will z-score the expression values for display.
This is sometimes called a standardized score.

```{r row_normalize}
# PLIER::rowNorm z-scores the gene (row) values

> zscores_mat <- PLIER::rowNorm(high_var_mat)

Reviewer: If we want to include an alternative that follows the Z score formula:
```
(high_var_mat - rowMeans(high_var_mat)) / matrixStats::rowSds(high_var_mat)
```

(don't try this with column z scores!)

Author: Oh I thought for sure the `sweep()`-based solution would be faster, but that is not the case (my guess is `matrixStats::rowSds()` and `rowMeans()` is why this is faster).

```

Let's look at the mean and standard deviation of values for a randomly selected gene prior to standardization.

```{r random_row_mean}
mean(high_var_mat[25, ])
```

```{r random_row_sd}
sd(high_var_mat[25, ])
```

And now those values after standardization.

```{r random_z_mean, live = TRUE}
mean(zscores_mat[25, ])
```

> This value isn't exact zero but it's very close!

Reviewer: ```suggestion
This value isn't exactly zero but it's very close!
```

```{r random_z_sd, live = TRUE}
sd(zscores_mat[25, ])
```

The standard deviation is now 1.

> That means all the variance values should be equal to one, too.

Reviewer: Unless math is broken.
```suggestion
That means all the variance values will be equal to one, too.
```

```{r summary_row_variance}
summary(matrixStats::rowVars(zscores_mat))
```

If you're filtering to high variance values, it's important to do that _prior_ to standardizing.

### Heatmap itself!

Okay, now we're ready to make a heatmap complete with annotation bars.

```{r make_heatmap}
Heatmap(zscores_mat,
        # The distance metric used for clustering the rows
        clustering_distance_rows = "pearson",
        # Linkage method for row clustering
        clustering_method_rows = "average",
        # Distance metric for columns
        clustering_distance_columns = "pearson",
        # Linkage for columns

>         clustering_method_columns = "average",

Reviewer: Are these the defaults? (Do they have to be specified?) Worth noting either way.

Author: I _think_ the defaults are Pearson with complete linkage but I'll double check and update this later. I believe this distance and linkage combination is the default in tools we use in the next notebook, so I thought I would keep it consistent.

Author: Default distance is Euclidean

        show_row_names = FALSE,
        show_column_names = FALSE,
        # Add annotation bars to the top of the 
        top_annotation = column_annotation,
        # This will be used as the label for the color bar
        # of the cells of the heatmap itself
        name = "z-score")
```

```{r save_heatmap_png, live = TRUE}

> png(filename = heatmap_file, 

Reviewer: We should briefly remind them why we have to make the plot twice, once for screen & once to save.

    width = 11, 
    height = 7, 
    units = "in", 
    res = 300)
Heatmap(zscores_mat,
        clustering_distance_rows = "pearson",
        clustering_method_rows = "average",
        clustering_distance_columns = "pearson",
        clustering_method_columns = "average",
        show_row_names = FALSE,
        show_column_names = FALSE,
        top_annotation = column_annotation,
        name = "z-score")
dev.off()
```

## Spotlight on medulloblastoma

The heatmap we made with all of the histologies under consideration suggested that there's some structure _within_ histologies.
Let's take a closer look at the medulloblastoma samples.

```{r mb_sample_ids, live = TRUE}
medulloblastoma_samples <- filtered_histologies_df %>%
  filter(short_histology == "Medulloblastoma") %>%
  pull(Kids_First_Biospecimen_ID)
```

Use all genes to create a matrix that only contains medulloblastoma samples.

```{r mb_mat, live = TRUE}
medulloblastoma_mat <- rnaseq_mat[, medulloblastoma_samples]
```

> We'll use Pearson correlation for our distance metric here again, but let's breakdown what goes into hierarchical clustering a bit here.

Reviewer: ```suggestion
We'll use Pearson correlation for our distance metric here again, but let's break down what goes into hierarchical clustering a bit here.
```

```{r calculate_cor}
# cor() will take the correlation between columns - we're interested in the 
# relationship between medulloblastoma samples, so we don't need to transpose
# the matrix before using cor()
mb_sample_correlation <- cor(medulloblastoma_mat,
                             use = "pairwise.complete.obs",
                             method = "pearson")
```

Hierarchical clustering is performed on dissimilarities, so we'll subtract the correlation values from 1.

> To use `hclust()` -- the built-in method for hierarchical clustering in R -- we need a dissimilarity structure.

Reviewer: I feel like these sentences need a bit more clarity. What we need is a distance matrix... how is this different from just the (1-cor) matrix?

Author: I'll talk about what `as.dist()` does explicitly.

```{r cor_as_dist, live = TRUE}
mb_sample_dist <- as.dist(1 - mb_sample_correlation)
```

Now we're ready to perform [hierarchical clustering](https://en.wikipedia.org/wiki/Hierarchical_clustering) with `hclust()`.
This is an _agglomerative_ method of clustering - each sample starts in its own cluster and then at each step of the algorithm the two most similar clusters are joined until there's only a single cluster left.

```{r mb_hclust, live = TRUE}
mb_sample_hclust <- hclust(mb_sample_dist, 
                           method = "average")
```

Here, `method = "average"` refers to the agglomeration or linkage method used during clustering. 
The method we're using is also known as [UPGMA (unweighted pair group method with arithmetic mean)](https://en.wikipedia.org/wiki/UPGMA) and is how we specify what formula to use for calculating inter-cluster distances during hierarchical clustering.

> The distance between two clusters is the average distance between all pairs of points in the two clusters.

Reviewer: I don't know if this is more clear.
```suggestion
The distance between two clusters is the average distance between all pairs of points having one point from each of the two clusters.
```

Author: Hmmm I'll take another pass at it with your suggestion in mind.

Reviewer: Yeah, I didn't love mine either. This is one of those things that is easy to envision and hard to write succinctly.
